packages/truffle-app/contracts/Giveaways.sol

```diff
@@ -1,80 +1,119 @@
+        bytes24 id,
+        uint256 startTime,
+        uint256 endTime,
+        uint8 nWinners
+    ) public onlyOwner {
+        Giveaway memory giveaway;
+        giveaway.id = id;
+        giveaway.startTime = startTime;
+        giveaway.endTime = endTime;
+        giveaway.nWinners = nWinners;
+        giveaways[id] = giveaway;
+    }
+
+    function addParticipant(
+        bytes24 giveawayId,
+        string memory participant
+    ) public onlyOwner {
+        Giveaway storage giveaway = giveaways[giveawayId];
+
+        require(giveaway.id != bytes24(0), 'Giveaway does not exist');
+        require(
+            block.timestamp >= giveaway.startTime &&
+                block.timestamp <= giveaway.endTime,
+            'Giveaway is not active'
+        );
 
-    require(giveaway.startTime <= block.timestamp, "Giveaway hasn't started yet");
-    require(giveaway.endTime <= block.timestamp, "Giveaway hasn't ended yet");
-    require(giveaway.participants.length >= giveaway.nWinners, "Not enough participants to choose a winner");
-        
-    address[] memory eligibleParticipants = new address[](giveaway.participants.length);
-    uint16 numberOfEligibleParticipants = 0;
+        // Check if the participant has already participated
+        for (uint i = 0; i < giveaway.participants.length; i++) {
+            if (
+                keccak256(abi.encodePacked(giveaway.participants[i])) ==
+                keccak256(abi.encodePacked(participant))
+            ) {
+                revert('Participant has already participated');
+            }
+        }
 
-    // add participants to eligible participants list
-    for (uint16 i = 0; i < giveaway.participants.length; i++) {
-      eligibleParticipants[numberOfEligibleParticipants] = giveaway.participants[i];
-      numberOfEligibleParticipants++;
+        // Add the participant to the participants array
+        giveaway.participants.push(participant);
     }
 
-    // generate winners for eligible participants
-    for (uint8 i = 0; i < giveaway.nWinners; i++) {
-      uint16 randomIndex = uint16(uint256(keccak256(abi.encodePacked(blockhash(block.number - 1), block.timestamp, i))) % numberOfEligibleParticipants);
-      giveaway.winners.push(eligibleParticipants[randomIndex]);
-            
-      // remove winner from eligible participants array
-      numberOfEligibleParticipants--;
-      eligibleParticipants[randomIndex] = eligibleParticipants[numberOfEligibleParticipants];
+    function generateWinners(bytes24 id) public onlyOwner {
+        Giveaway storage giveaway = giveaways[id];
+
+        require(
+            giveaway.startTime <= block.timestamp,
+            "Giveaway hasn't started yet"
+        );
+        require(
+            giveaway.endTime <= block.timestamp,
+            "Giveaway hasn't ended yet"
+        );
+        require(
+            giveaway.participants.length >= giveaway.nWinners,
+            'Not enough participants to choose a winner'
+        );
```

Comment:
In this case shouldn't the winners be all of the participants?

Reply:
I feel like the number of winners should be the minimum requirement in terms of the minimum number of participants for a giveaway to be considered valid, but that is a matter of preference